### ISSUE TITLE
Halide-generator module?

### ISSUE THREAD

**Issue by antonysigma** [mesonbuild/meson#10687]
Following the discussion at https://github.com/halide/Halide/issues/6688#issuecomment-1204764086 , I am curious about how we the developer can contribute an (unstable) module to support a halide-generator macro, to implement the CMake macro at https://github.com/halide/Halide/blob/main/README_cmake.md#rungenmain

Quoting the original discussion:

> Sorry, Meson is an absolute non-starter for Halide as it has no way to abstract and redistribute rules for calling Halide generators. This is not likely to change as sub-Turing-completeness is one of Meson's core design principles.

> The solution that makes the most sense to me is for Halide to install halide-generator.pc, halide-runtime.pc, etc. as pkg-config files, which are very simple to create and install from cmake, and then Meson/autotools can detect those pretty easily if you install Halide to the system. 

Personally, I don't mind manually expanding the CMake macro into Meson `custom_target`, in accordance to the [meson's guideline](https://mesonbuild.com/Syntax.html#userdefined-functions-and-methods). I even have a working implementation with `foreach`. But witnessing the debate between meson teams and halide teams regarding macros, I wish to steer the discussion to the idea of an `unstable-halide` module in Meson.

**The context** The need arises in the halide toolchain for scientific computing, when the (algorithm) developer synthesizes an optimized algorithm on the CPU. Halide implement this with the Canadian-cross pattern:

1. The `algorithm.cpp` compiles and links with `libHalide.so` on the host machine, emitting a `generator` program;
2. The `generator` program runs on the build machine, emitting either an object file `algorithm.o` or an assembly file `algorithm.s`;
3. The generated object file `algorithm.o` links with the user's main program `main.cpp` and the halide header-only runtime library `halide-runtime.hpp`, emitting a `main` program on the target machine.

**The current solution (CMake)** Halide developers contributed a [CMakeGeneratorHelper.cmake](https://github.com/halide/Halide/blob/main/cmake/HalideGeneratorHelpers.cmake), which contains a macro to generate the CMake target `algorithm.o` from user-provided `algorithm.cpp` input file. The intermediate build stages are hidden from the user.

**The current solution (Meson)** I followed the Meson's guidelines to explicitly expanded Steps 1 to 3 into `executable('generator')`, `custom_target('algorithm.o')`, and `executable('main')`. Please refer to the code example in the link below:

https://github.com/comp-imaging/ProxImaL/blob/master/proximal/halide/meson.build#L141-L169

**The question** Since the pattern "steps 1 to 3" is so commonly used by the halide toolchain, is it feasible for developers to contribute an `unstable-halide` module to Meson codebase?

**Comment by dcbaker** [comment]
If someone is willing to write and maintain it, then I don’t see any reason why not

**Comment by antonysigma** [comment]
Thanks. I guess we all agreed a Halide (unstable) module is not sustainable if no one maintains it.

Hi @eli-schwartz : I created this thread to address your earlier suggestions to import a user-defined macro/function from `halide-generator.pc`. I took a closer look at the Halide generator CMake rules, and I figured what it contains is equivalent to a "module" in Meson, not a simple `custom_target`.
